### ISSUE TITLE
Bug on Monads

### ISSUE THREAD

**Issue by vicdoval** [nortikin/sverchok#1860]
First thank you all for your work, I love Sverchok, its such a great tool its hard to explain.

I have detected a strange behavior in the "Monad":
  -Inside a group create a "Vector Math Node"
   -Connect it to the "inputs node"
You will see that it should create intermediately another "empty input" but it does not.
After you try to move another wire inside the Monad the new "empty input" its created, but also the last input is half created again (you can only see it from outside the node)

I hope you got it and it helped.

Thanks again for your work!

**Comment by zeffii** [comment]
yep, can confirm
```python
Traceback (most recent call last):
  File "sverchok\nodes\scene\monad.py", line 91, in update
    new_socket.prop_name = prop_name
TypeError: bpy_struct: item.attr = val: VerticesSocket.prop_name expected a string type, not NoneType
```

**Comment by zeffii** [comment]
can't promise a rapid turnaround if the first few ideas don't work :)  bear with us.

**Comment by zeffii** [comment]
worth trying this branch: https://github.com/nortikin/sverchok/pull/1861

**Comment by vicdoval** [comment]
It works. 👍  Thanks again :) 
There is this minor bug in the UI when you plug the "Vector Math" the "vector drop down menu" appears on the Monad UI but it is blank and non-editable.  Probably it shouldn't appear on there. Anyway probably it is not very important.  :)

**Comment by zeffii** [comment]
it isnt fixed yet. Optimistically this'll take a day. But there are other updates for sockets that can solve some of the complexities of monads inner workings. So failing a quickfix, a more elaborate one is on the horizon

**Comment by zeffii** [comment]
@vicdoval  also locally you could try 
```python
new_socket.prop_name = prop_name or ''
```
that way you will get the socket only, and not the dropdown (which isn't working -- i've not had time to look into it yet)

**Comment by zeffii** [comment]
i don't have much time for this right this minute but can write a few notes for later, or for the interested reader.

~If i recall correctly:~  this is not what happens... but ideally it would be  

Currently each time you make a monad, or update a monad (by adding sockets etc) it 
   - generates a new node class dynamically 
   - the new node class has replicas of `bpy.props.**Property` required by the sockets that made the input/output
   - then registers it

I thought FloatVectorProperty were registered, but it seems not the case anymore. There have been some registration changes in Blender in recent 2.79 builds. You can confirm this by trying a 2.78 build with the same Vector Math node. I don't recall leaving that broken -- but in honesty i might have. which is weird seen as that's the node i think i used to convince myself that Monad was less broken than before. ( if you go thru commits and images of the most recent sprint of monad development i'd be surprised if you didn't find screenshots of Vector Math node passing it's `v3_value_0` socket property to the monad class. )

It's definitely simpler to not track the `prop_name` for vector type sockets. By not doing that - the extra work is negligible while the monad can still be used effectively by connecting a socket to the outside of the monad.

dynamically generating properties isn't hard, and even avoiding property name collisions is manageable, but it is a lot of extra complexity to maintain, and for very little benefit.

One thing someone with a bit of C knowledge could help us out with is to add the ability to pass a callback to the bpy.types float/int/vector so the min / max can be function and defined at runtime rather than the static way it is now. ideasman showed be a while back where to do this, but it's a little out of my development environment capacity (and time) to stare at it until it makes sense. so if anyone feels up for it.... that would be great. and would be awesome for AN too.

**Comment by vicdoval** [comment]
I tried your last suggestion :

new_socket.prop_name = prop_name or ''

It works very well and does not look broken  :D
I also tried the actual "master" with 2.78 and the bug was already there.

(No C knowledge hosted on me. Sorry I cant help on the min/max callbacks.)


I think this solution is pretty neat and I would consider this Issue fixed. The next step is a nice enhancement but not so necessary on my view.

Thanks!
